pkg/framework/profile/profile.go

```diff
@@ -0,0 +1,310 @@
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package profile
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"sigs.k8s.io/descheduler/metrics"
+	"sigs.k8s.io/descheduler/pkg/api"
+	"sigs.k8s.io/descheduler/pkg/descheduler/evictions"
+	podutil "sigs.k8s.io/descheduler/pkg/descheduler/pod"
+	"sigs.k8s.io/descheduler/pkg/framework"
+	"sigs.k8s.io/descheduler/pkg/framework/pluginregistry"
+	"sigs.k8s.io/descheduler/pkg/framework/plugins/defaultevictor"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/informers"
+	clientset "k8s.io/client-go/kubernetes"
+
+	"k8s.io/klog/v2"
+)
+
+// evictorImpl implements the Evictor interface so plugins
+// can evict a pod without importing a specific pod evictor
+type evictorImpl struct {
+	podEvictor    *evictions.PodEvictor
+	evictorFilter framework.EvictorPlugin
+}
+
+var _ framework.Evictor = &evictorImpl{}
+
+// Filter checks if a pod can be evicted
+func (ei *evictorImpl) Filter(pod *v1.Pod) bool {
+	return ei.evictorFilter.Filter(pod)
+}
+
+// PreEvictionFilter checks if pod can be evicted right before eviction
+func (ei *evictorImpl) PreEvictionFilter(pod *v1.Pod) bool {
+	return ei.evictorFilter.PreEvictionFilter(pod)
+}
+
+// Evict evicts a pod (no pre-check performed)
+func (ei *evictorImpl) Evict(ctx context.Context, pod *v1.Pod, opts evictions.EvictOptions) bool {
+	return ei.podEvictor.EvictPod(ctx, pod, opts)
+}
+
+func (ei *evictorImpl) NodeLimitExceeded(node *v1.Node) bool {
+	return ei.podEvictor.NodeLimitExceeded(node)
+}
+
+// handleImpl implements the framework handle which gets passed to plugins
+type handleImpl struct {
+	clientSet                 clientset.Interface
+	getPodsAssignedToNodeFunc podutil.GetPodsAssignedToNodeFunc
+	sharedInformerFactory     informers.SharedInformerFactory
+	evictor                   *evictorImpl
+}
+
+var _ framework.Handle = &handleImpl{}
+
+// ClientSet retrieves kube client set
+func (hi *handleImpl) ClientSet() clientset.Interface {
+	return hi.clientSet
+}
+
+// GetPodsAssignedToNodeFunc retrieves GetPodsAssignedToNodeFunc implementation
+func (hi *handleImpl) GetPodsAssignedToNodeFunc() podutil.GetPodsAssignedToNodeFunc {
+	return hi.getPodsAssignedToNodeFunc
+}
+
+// SharedInformerFactory retrieves shared informer factory
+func (hi *handleImpl) SharedInformerFactory() informers.SharedInformerFactory {
+	return hi.sharedInformerFactory
+}
+
+// Evictor retrieves evictor so plugins can filter and evict pods
+func (hi *handleImpl) Evictor() framework.Evictor {
+	return hi.evictor
+}
+
+type profileImpl struct {
```

Comment:
Cant we use the profile that we already have in our api? I think it is a bit confusing:

- we read a config file
- we populate our Deschedule policy that in fact has a profile (that in the end has plugins inside).

then we

- move things from the profile of the api into an internal profile that has Run[*]Plugins methods to be able to call those for each profile.

Can't we just the profile that we have in the api? And, if it doesn't deliver something that we need, can't we just change it (so we only have one type with this name inside the code)?

Reply:
Do you mean https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/api/types.go#L60? If so, `profileImpl` is actual implementation of handling a single profile configuration. We tend to keep the api definition from anything handling the api. Though, maybe I just misunderstand.

Reply:
Yes, I mean that one. It is not really that separate, if we think that when we change anything in the api's profile, we would probably also change somethings in this profileImpl. 

Maybe we could rename it in this new place? configuredProfile? It is also about the methods and instantiations. 

maybe [here](https://github.com/ingvagabund/descheduler/blob/2ae5ec8ac37770b5310d9ab8f8bd3e1394d07a32/pkg/descheduler/descheduler.go#L305) prfl could be named configuredProfile, or just profileImpl for the instance as well. 

maybe [methods](https://github.com/ingvagabund/descheduler/blob/2ae5ec8ac37770b5310d9ab8f8bd3e1394d07a32/pkg/framework/profile/profile.go#L167) could also be different so it is also clear this is a different profile, maybe NewConfiguredProfile or something like that

I would also change the name of the package and folder I think.

Just so it is clear that this is a different profile then, that this is a implementation of handling a single profile configuration. I am not sure about the names though. I would not block the PR because of this also

Reply:
Good points. The upstream kube-scheduler code is using `KubeSchedulerProfile` name in the API part: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-scheduler/config/v1/types.go#L132. We could use the same naming schema and rename our Profile into DeschedulerProfile. While keeping the rest as it is?

Reply:
Yeah, that also works I think